# Arm B: Clinically Defined Subgroups (Sex Stratified)

This notebook implements Arm B of the study. Patients are split into two clinically defined subgroups by sex, and each classifier is tuned and trained separately within each subgroup. Arm B answers RQ1 (does clinical stratification change performance compared with the global model of Arm A?) and, together with Arm C, RQ3.

Sex is the stratification variable because cardiovascular disease can differ between males and females in risk factor profiles and in how features relate to the outcome, and sex specific models have been reported to differ in performance and in their important predictors (proposal, Section 3.3). As a binary variable it also produces only two subgroups, which keeps each one as large as possible given 297 patients.

Everything except the grouping is held identical to Arm A by importing it from `study_config.py`: the dataset, the 9 features, the preprocessing, the three classifiers and their grids, the inner tuning procedure, the outer folds (`fold_id.csv`, created by Arm A) and the 0.5 threshold. The one deliberate difference is that `sex` is removed from the feature matrix, explained in Section 4.

## 1. Imports and configuration

All settings come from `study_config.py`, so they cannot differ from Arm A's. `K_INNER = 10` also works within the female subgroup: each female training set has about 86 patients and about 22 with disease, enough for a stratified 10 fold inner split.

In [1]:
import hashlib
import json
import os

import numpy as np
import pandas as pd

from sklearn.model_selection import StratifiedKFold, GridSearchCV
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score

import study_config as cfg
from study_config import RANDOM_STATE, K_OUTER, K_INNER, THRESHOLD, FEATURE_SET, LOGREG

cfg.make_dirs()
STRATIFIER = "sex"
SEX_GROUPS = {"male": 1.0, "female": 0.0}
print(f"FEATURE_SET={FEATURE_SET!r}, K_OUTER={K_OUTER}, K_INNER={K_INNER}, "
      f"THRESHOLD={THRESHOLD}, RANDOM_STATE={RANDOM_STATE}, tuning={cfg.TUNING_SCORING!r}")

FEATURE_SET='selected', K_OUTER=10, K_INNER=10, THRESHOLD=0.5, RANDOM_STATE=2, tuning='roc_auc'


## 2. Load the dataset and build the target

Arm B uses the same cleaned file as Arm A and the same binary target (`num > 0` means disease present).

In [2]:
df = pd.read_csv(cfg.CLEAN_DATA_FILE)
assert df.shape[0] == 297, "Arm B expects the same 297 patient dataset as Arm A."
y = (df["num"] > 0).astype(int)
sex = df["sex"].to_numpy()
print("Loaded shape:", df.shape, "| positive rate:", round(y.mean(), 3))

Loaded shape: (297, 15) | positive rate: 0.461


## 3. Subgroup sizes

Both subgroups must be large enough, and contain enough patients of each class, for nested cross validation to be well defined within them (proposal, Section 3.2). Males outnumber females about two to one, and the disease rate differs sharply between them (about 55% of males vs 26% of females). Patients are not removed to equalise the subgroups.

In [3]:
sex_label = df["sex"].map({v: k for k, v in SEX_GROUPS.items()})
summary = pd.crosstab(sex_label, y.map({0: "no disease", 1: "disease"}), margins=True, margins_name="total")
summary["disease_rate"] = (summary["disease"] / summary["total"]).round(3)
summary

num,disease,no disease,total,disease_rate
sex,,,,
female,25,71,96,0.260
male,112,89,201,0.557
total,137,160,297,0.461


## 4. Features

Arm B starts from the same 9 features as Arm A and removes only `sex`, leaving 8. Once patients are split by sex, `sex` takes a single value within each subgroup, so it carries no information for that subgroup's model. The information is not lost: it is used to decide which model a patient goes to. Leaving the constant column in would not help and can slightly hurt the random forest, which samples a random subset of columns at each split and would sometimes waste a draw on a column that can never split.

The check below confirms `sex` really is constant within each subgroup rather than assuming it.

In [4]:
EXCLUDE = (STRATIFIER,)
X = df[cfg.get_feature_list(FEATURE_SET, exclude=EXCLUDE)]
for g, code in SEX_GROUPS.items():
    assert df.loc[sex == code, STRATIFIER].nunique() == 1
print(f"Features used ({X.shape[1]}): {list(X.columns)}")
n_cols = cfg.get_preprocessor(FEATURE_SET, exclude=EXCLUDE).fit_transform(X).shape[1]
print(f"8 raw features -> {n_cols} model columns (Arm A has 13; the only missing column is sex)")
assert n_cols == 12

Features used (8): ['age', 'thalach', 'oldpeak', 'ca', 'cp', 'slope', 'thal', 'exang']
8 raw features -> 12 model columns (Arm A has 13; the only missing column is sex)


## 5. Reuse Arm A's outer folds

Arm B loads `fold_id.csv` from Arm A, so every patient is held out in exactly the same fold in every arm. Before using it, the notebook checks that the file's MD5 hash and the configuration recorded in `run_manifest.json` match this notebook's own settings. If anything differs, it stops rather than producing results on a different partition.

Within each outer fold, a patient's subgroup training set is "training fold and same sex", and their subgroup validation set is "validation fold and same sex". The table below shows that every fold gives both subgroups enough training patients of each class for the inner 10 fold split.

In [5]:
fold_id = pd.read_csv(cfg.FOLD_FILE)["fold"].to_numpy()
with open(cfg.MANIFEST_FILE) as f:
    manifest = json.load(f)
assert hashlib.md5(open(cfg.FOLD_FILE, "rb").read()).hexdigest() == manifest["fold_id_md5"], \
    "fold_id.csv changed since Arm A ran; rerun Arm A, then this notebook."
for key, value in {"feature_set": FEATURE_SET, "k_outer": K_OUTER, "k_inner": K_INNER,
                   "random_state": RANDOM_STATE, "threshold": THRESHOLD,
                   "tuning_scoring": cfg.TUNING_SCORING}.items():
    assert manifest[key] == value, f"run_manifest.json {key}={manifest[key]!r} but this notebook uses {value!r}"
assert len(fold_id) == len(df) and set(np.unique(fold_id)) == set(range(K_OUTER))
print("fold_id.csv and run_manifest.json verified against study_config.")

rows = []
for k in range(K_OUTER):
    for g, code in SEX_GROUPS.items():
        tr = (fold_id != k) & (sex == code)
        rows.append({"fold": k, "subgroup": g, "train_n": int(tr.sum()),
                     "train_disease": int(y[tr].sum()), "train_no_disease": int((1 - y[tr]).sum()),
                     "validation_n": int(((fold_id == k) & (sex == code)).sum())})
fold_sizes = pd.DataFrame(rows)
assert fold_sizes[["train_disease", "train_no_disease"]].min().min() >= K_INNER
fold_sizes.pivot(index="fold", columns="subgroup", values=["train_n", "train_disease", "validation_n"])

fold_id.csv and run_manifest.json verified against study_config.


train_n      train_disease      validation_n     
subgroup  female male        female male       female male
fold                                                      
0             87  180            23  100            9   21
1             86  181            23  101           10   20
2             86  181            22  101           10   20
3             86  181            22  101           10   20
4             86  181            22  101           10   20
5             86  181            22  101           10   20
6             86  181            22  101           10   20
7             87  181            23  101            9   20
8             87  181            23  101            9   20
9             87  181            23  100            9   20

## 6. Nested cross validation within each subgroup

For each classifier and outer fold, a separate model is tuned and trained for each sex. Hyperparameters are chosen by `GridSearchCV` with an inner stratified 10 fold split of that subgroup's training patients only, using the same grid and ROC AUC criterion as Arm A; the outer validation fold is never seen during tuning. The male and female models are completely independent: neither ever sees the other sex's patients. Because the subgroups differ, they may settle on different hyperparameters. That is expected, since both choose from the same candidates by the same rule.

Each validation patient is scored by their own sex's model and classed with the shared 0.5 threshold. The male and female predictions are then **pooled** into one set covering the whole validation fold before accuracy, F1 and ROC AUC are computed. This pooled, population level score is the primary Arm B result, because it measures the same thing as Arm A's score: performance on all patients in the fold (proposal, Section 3.5).

**ROC AUC caveat.** The pooled ROC AUC ranks male and female patients together using probabilities from two separate models, so it partly reflects how those two models' probability scales line up, not only how well each one ranks its own patients. Accuracy and F1 are not affected, because the threshold is applied within each subgroup before pooling. As a check, `roc_auc_within_group` records the patient weighted average of the male and female AUCs computed separately. `cross_arm_comparison.ipynb` reports both.

In [6]:
models = cfg.get_models(FEATURE_SET, exclude=EXCLUDE)

fold_rows, group_rows, oof_rows, best_param_rows = [], [], [], []
patient_id = df.index.to_numpy()

for name, (pipe, grid) in models.items():
    for k in range(K_OUTER):
        train, validation = fold_id != k, fold_id == k
        pooled_true, pooled_proba, pooled_pred, group_auc, group_n = [], [], [], [], []

        for g, code in SEX_GROUPS.items():
            g_train, g_val = train & (sex == code), validation & (sex == code)

            inner = StratifiedKFold(n_splits=K_INNER, shuffle=True, random_state=RANDOM_STATE)
            search = GridSearchCV(pipe, grid, cv=inner, scoring=cfg.TUNING_SCORING, n_jobs=-1)
            search.fit(X[g_train], y[g_train])

            proba = search.best_estimator_.predict_proba(X[g_val])[:, 1]
            pred = (proba >= THRESHOLD).astype(int)
            y_val = y[g_val].to_numpy()
            auc = roc_auc_score(y_val, proba) if len(np.unique(y_val)) > 1 else np.nan

            group_rows.append({"model": name, "fold": k, "subgroup": g, "n": len(y_val),
                               "accuracy": accuracy_score(y_val, pred),
                               "f1": f1_score(y_val, pred, zero_division=0), "roc_auc": auc})
            best_param_rows.append({"model": name, "fold": k, "subgroup": g, "best_params": search.best_params_})
            group_auc.append(auc)
            group_n.append(len(y_val))
            for pid, yt, p, c in zip(patient_id[g_val], y_val, proba, pred):
                oof_rows.append({"patient_id": int(pid), "fold": k, "subgroup": g, "model": name,
                                 "y_true": int(yt), "proba": float(p), "pred": int(c)})
            pooled_true.append(y_val)
            pooled_proba.append(proba)
            pooled_pred.append(pred)

        yt_all, pr_all, pd_all = map(np.concatenate, (pooled_true, pooled_proba, pooled_pred))
        aucs, ns = np.array(group_auc), np.array(group_n, dtype=float)
        ok = ~np.isnan(aucs)
        fold_rows.append({
            "model": name, "fold": k,
            "accuracy": accuracy_score(yt_all, pd_all),
            "f1": f1_score(yt_all, pd_all, zero_division=0),
            "roc_auc": roc_auc_score(yt_all, pr_all),
            "roc_auc_within_group": float(np.average(aucs[ok], weights=ns[ok])) if ok.any() else np.nan,
        })

fold_results_df = pd.DataFrame(fold_rows)
group_results_df = pd.DataFrame(group_rows)
oof_df = pd.DataFrame(oof_rows)
best_params_df = pd.DataFrame(best_param_rows)
best_params_df.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armB_best_params.csv"), index=False)
print("Nested cross validation complete:", len(fold_results_df), "classifier x fold rows,",
      len(group_results_df), "classifier x fold x subgroup rows")

Nested cross validation complete: 30 classifier x fold rows, 60 classifier x fold x subgroup rows


## 7. Results

The files saved in `results/` match Arm A's layout so the arms can be compared directly:

* `results/armB_fold_results.csv`: pooled accuracy, F1 and ROC AUC per classifier and fold, plus `roc_auc_within_group`.
* `results/armB_results.csv`: mean and standard deviation across the 10 folds of the pooled metrics.
* `results/predictions/armB_predictions.csv`: one out of fold prediction per patient per classifier, with the patient's subgroup.

Per subgroup scores (a secondary, descriptive breakdown) and the chosen hyperparameters are saved in `diagnostics/`. Confusion matrices and tests between arms are in `cross_arm_comparison.ipynb`.

In [7]:
assert oof_df.groupby("model")["patient_id"].nunique().eq(len(df)).all()
assert oof_df.groupby(["model", "patient_id"]).size().max() == 1

fold_results_df.to_csv(os.path.join(cfg.RESULTS_DIR, "armB_fold_results.csv"), index=False)
oof_df.to_csv(os.path.join(cfg.PREDICTIONS_DIR, "armB_predictions.csv"), index=False)

results = fold_results_df.groupby("model", sort=False)[["accuracy", "f1", "roc_auc"]].agg(["mean", "std"])
results.columns = [f"{m}_{s}" for m, s in results.columns]
results = results.reset_index()
results.to_csv(os.path.join(cfg.RESULTS_DIR, "armB_results.csv"), index=False)
results.set_index("model").round(3)

,accuracy_mean,accuracy_std,f1_mean,f1_std,roc_auc_mean,roc_auc_std
model,,,,,,
Logistic regression,0.832,0.080,0.813,0.089,0.890,0.065
Random forest,0.839,0.066,0.821,0.073,0.927,0.042
Support vector machine,0.815,0.074,0.791,0.090,0.896,0.049


### Per subgroup scores (secondary, descriptive)

Mean and standard deviation across folds of each subgroup's own scores. These show where any difference from Arm A comes from; they are not the basis for answering RQ1. Each female validation fold holds only about 10 patients with 2 or 3 disease cases, so the female scores move a lot from fold to fold by construction.

In [8]:
group_summary = group_results_df.groupby(["subgroup", "model"], sort=False)[["accuracy", "f1", "roc_auc"]].agg(["mean", "std"])
group_summary.columns = [f"{m}_{s}" for m, s in group_summary.columns]
group_summary = group_summary.reset_index()
group_summary.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armB_subgroup_results.csv"), index=False)
group_summary.set_index(["subgroup", "model"]).round(3)

,,accuracy_mean,accuracy_std,f1_mean,f1_std,roc_auc_mean,roc_auc_std
subgroup,model,,,,,,
male,Logistic regression,0.811,0.096,0.818,0.101,0.891,0.065
female,Logistic regression,0.873,0.101,0.760,0.200,0.917,0.114
male,Random forest,0.812,0.098,0.832,0.086,0.910,0.060
female,Random forest,0.894,0.071,0.690,0.287,0.905,0.131
male,Support vector machine,0.777,0.100,0.791,0.103,0.881,0.067
female,Support vector machine,0.897,0.067,0.776,0.157,0.871,0.091


## 8. Female logistic regression: threshold check

Penalised logistic regression fitted on a small subgroup with a low disease rate can be pushed towards very strong regularisation. Strong regularisation shrinks every coefficient towards zero, so all predicted probabilities move towards the subgroup's base rate (about 26% for females). The model may still rank patients correctly (good ROC AUC) while rarely or never crossing the 0.5 threshold, which would show up as poor sensitivity and F1 for reasons unrelated to how well it discriminates.

The table checks this per fold: the chosen `C` (smaller means stronger regularisation), the highest predicted probability among female validation patients, the female training disease rate, and the female ROC AUC. A fold where the maximum probability is below 0.5 would predict no female patient as diseased.

In [9]:
f_lr = (oof_df["model"] == LOGREG) & (oof_df["subgroup"] == "female")
bp = best_params_df[(best_params_df["model"] == LOGREG) & (best_params_df["subgroup"] == "female")].set_index("fold")
female_check = pd.DataFrame({
    "chosen_C": bp["best_params"].map(lambda p: float(p["clf__C"])),
    "max_proba_female_val": oof_df[f_lr].groupby("fold")["proba"].max(),
    "female_train_disease_rate": pd.Series({k: y[(fold_id != k) & (sex == 0.0)].mean() for k in range(K_OUTER)}),
    "female_roc_auc": group_results_df[(group_results_df["model"] == LOGREG)
                                       & (group_results_df["subgroup"] == "female")].set_index("fold")["roc_auc"],
}).round(3)
female_check.index.name = "fold"
female_check.to_csv(os.path.join(cfg.DIAGNOSTICS_DIR, "armB_female_logreg_check.csv"))
female_check

,chosen_C,max_proba_female_val,female_train_disease_rate,female_roc_auc
fold,,,,
0,0.001,0.528,0.264,0.643
1,0.167,0.883,0.267,1.000
2,0.013,0.739,0.256,1.000
3,0.167,0.984,0.256,1.000
4,0.167,0.965,0.256,0.905
5,0.001,0.536,0.256,1.000
6,0.013,0.668,0.256,0.905
7,0.599,0.765,0.264,1.000
8,0.599,0.987,0.264,0.857


**Reading the threshold check.** No fold collapses completely: every fold's highest female probability is above 0.5. Folds 0 and 5 come close. In both, the inner search chose the strongest regularisation on the grid (`C = 0.001`), and the highest female probability was only 0.528 and 0.536. Fold 5 shows the mechanism clearly: its female ROC AUC was 1.00, so the model ranked the female patients perfectly while its probabilities were squeezed towards the base rate, only just crossing the threshold. Fold 0 also had the lowest female ROC AUC (0.64).

The chosen `C` is small (strong regularisation) in most folds, between 0.001 and 0.6. Each female inner validation split holds about 9 patients with only 2 or 3 disease cases, so the inner ROC AUC that drives this choice is very noisy. This is a direct cost of stratifying a small sample: tuning a separate model on 86 patients is far less stable than tuning one model on 267.

## 9. Notes

* **Pooled results.** Logistic regression: accuracy 0.832, F1 0.813, ROC AUC 0.890. Random forest: 0.839, 0.821, 0.927. Support vector machine: 0.815, 0.791, 0.896 (means across 10 folds). Whether these differ from Arm A is tested in `cross_arm_comparison.ipynb`, using the same folds, rather than judged from these numbers alone.
* **Per subgroup.** Female ROC AUC is high (0.87 to 0.92) but female F1 is lower and much less stable (0.69 to 0.78, standard deviation up to 0.29). With a female disease rate of 26% and only 2 or 3 disease cases per female validation fold, missing a single case moves F1 a long way; the threshold check above shows that ranking quality is not the main problem.
* **Controlled comparison.** Compared with Arm A, the only differences are that each sex gets its own tuned model and that `sex` is removed from the features (it is constant within a subgroup). Data, the other 8 features, preprocessing, classifiers, grids, tuning, folds and threshold all come from `study_config.py` and `fold_id.csv`, and are checked against `run_manifest.json` before the models run.